clusterloader2/pkg/measurement/common/network/worker_pod_info.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package network captures network performance metrics
+// for protocol TCP,UDP,HTTP etc. The metrics are collected for baseline (1:1),
+// scale (N:M) pod ratios.Client and server pods located on different worker
+// nodes exchange traffic for specified time to measure the performance metrics.
+package network
+
+import (
+	"container/heap"
+	"fmt"
+
+	"k8s.io/klog"
+)
+
+// workerNodeInfo represents node in heap
+type workerNodeInfo struct {
+	nodeName     string
+	numberOfPods int
+}
+
+type nodeHeap []*workerNodeInfo
+
+func (nh nodeHeap) Len() int { return len(nh) }
+func (nh nodeHeap) Less(i, j int) bool {
+	return nh[i].numberOfPods > nh[j].numberOfPods
+}
+
+func (nh nodeHeap) Swap(i, j int) {
+	nh[i], nh[j] = nh[j], nh[i]
+}
+
+func (nh *nodeHeap) Push(x interface{}) {
+	*nh = append(*nh, x.(*workerNodeInfo))
+}
+
+func (nh *nodeHeap) Pop() interface{} {
+	old := *nh
+	n := len(old)
+	x := old[n-1]
+	*nh = old[0 : n-1]
+	return x
+}
+
+type podList []workerPodData
+
+type workerPodInfo struct {
+	workerPodMap map[string]podList
+}
+
+func (l *podList) Pop() (*workerPodData, error) {
+	if len(*l) == 0 {
+		return nil, fmt.Errorf("empty pod list")
+	}
+	lastElement := (*l)[len(*l)-1]
+	*l = (*l)[:len(*l)-1]
+	return &lastElement, nil
+}
+
+func newUniquePodPair(index int, sourcePodName, sourcePodIP, destinationPodName, destinationPodIP string) uniquePodPair {
+	return uniquePodPair{
+		pairIndex:          index,
+		sourcePodName:      sourcePodName,
+		sourcePodIP:        sourcePodIP,
+		destinationPodName: destinationPodName,
+		destinationPodIP:   destinationPodIP}
+}
+
+func (wp *workerPodInfo) formUniquePodPair() []uniquePodPair {
+	nh := make(nodeHeap, 0, len(wp.workerPodMap))
+	podPairList := make([]uniquePodPair, 0)
+	for key, value := range wp.workerPodMap {
+		nh = append(nh, &workerNodeInfo{nodeName: key, numberOfPods: len(value)})
+	}
+	heap.Init(&nh)
+	for i := 0; nh.Len() > 0; i++ {
+		firstNode := heap.Pop(&nh).(*workerNodeInfo)
+		secondNode := heap.Pop(&nh).(*workerNodeInfo)
```

Comment:
This can potentially fail if there was just 1 element on the heap before.

You need to replace with `nh.Len() > 1`

Reply:
modified